From 88d35985828f28b4b10127b8057cd1db4171fc0d Mon Sep 17 00:00:00 2001
From: Petter Reinholdtsen <pere@hungry.com>
Date: Fri, 30 May 2014 10:09:57 +0200
Subject: [PATCH] Make it possible to link with the system libmpeg3 library.
 Add new configure option --with-external-libmpeg3 to not link with the
 internal libmpeg3 version.  This require at least libmpeg3 version 1.7 or
 newer.  Only handle header for external library in <mpeg3/libmpeg3.h>, and
 with a pkg-config setup available.  This should work with Debian Jessie and
 Fedora.

Based on patch from Kwizart and the RPM build of Cinelerra.
---
 Makefile.am           | 11 +++++++++--
 cinelerra/Makefile.am |  4 ++--
 configure.ac          | 47 +++++++++++++++++++++++++++++++++++++++++++++--
 mpeg2enc/Makefile.am  |  2 +-
 mplexlo/Makefile.am   |  4 ++--
 quicktime/Makefile.am |  4 ++--
 6 files changed, 61 insertions(+), 11 deletions(-)

diff --git a/Makefile.am b/Makefile.am
index 5bef08f..71648f3 100644
--- a/Makefile.am
+++ b/Makefile.am
@@ -1,5 +1,12 @@
-SUBDIRS=libmpeg3 quicktime mpeg2enc toolame-02l \
-	guicast cinelerra mplexlo plugins po image m4
+if HAVE_LIBMPEG3_INTERNAL
+  MAYBE_LIBMPEG3 = libmpeg3
+endif
+if HAVE_MPEG3IO_EOF
+  MAYBE_MPLEXLO = mplexlo
+endif
+
+SUBDIRS=$(MAYBE_LIBMPEG3) $(MAYBE_MPLEXLO) mpeg2enc toolame-02l \
+	quicktime guicast cinelerra plugins po image m4
 EXTRA_DIST = admin debian doc depcomp README.BUILD LICENSE \
 	cinelerra-cvs-current.spec
 AUTOMAKE_OPTIONS=foreign
diff --git a/cinelerra/Makefile.am b/cinelerra/Makefile.am
index 6ddb572..5ad0853 100644
--- a/cinelerra/Makefile.am
+++ b/cinelerra/Makefile.am
@@ -337,7 +337,7 @@ cinelerra_SOURCES = aattachmentpoint.C \
 AM_CPPFLAGS = \
 	-I$(top_srcdir)/guicast \
 	-I$(top_srcdir)/quicktime \
-	-I$(top_srcdir)/libmpeg3
+	$(LIBMPEG3_CFLAGS)
 
 AM_CXXFLAGS = \
 	$(LARGEFILE_CFLAGS) \
@@ -673,7 +673,7 @@ cinelerra_LDADD = \
 	$(top_builddir)/toolame-02l/libtoolame.la \
 	$(top_builddir)/quicktime/libquicktimehv.la \
 	$(top_builddir)/guicast/libguicast.la \
-	$(top_builddir)/libmpeg3/libmpeg3hv.la \
+	$(LIBMPEG3_LIBS) \
 	@FFMPEG_LIBS@ \
 	$(XIPH_LIBS) \
 	$(A52DEC_LIBS) \
diff --git a/configure.ac b/configure.ac
index 7e0206c..8c86c32 100644
--- a/configure.ac
+++ b/configure.ac
@@ -343,13 +343,56 @@ rm -f cinelerra/versioninfo.h > /dev/null 2>&1
 
 ############# END BUILDINFO display, (for displaying version / date)
 
-############# CSS SUPPORT IN LIBMPEG3
+############ external libmpeg3
+AC_ARG_WITH([external-libmpeg3], AC_HELP_STRING([--with-external-libmpeg3], [use external libmpeg3 library]))
+
+if test "x$with_external_libmpeg3" = "xyes"; then
+	PKG_CHECK_MODULES(LIBMPEG3, libmpeg3,[libmpeg3_system=yes],:)
+	AC_CHECK_HEADER(mpeg3/libmpeg3.h)
+	# the mpeg3io_eof macro in mpeg3protos.h is used by mplexlo,
+	# make sure it is available
+	old_LIBS="$LIBS"
+	LIBS="$LIBS -lmpeg3"
+	AC_LINK_IFELSE([AC_LANG_PROGRAM(
+[[#include <mpeg3/libmpeg3.h>
+#include <mpeg3/mpeg3protos.h>
+void test() { mpeg3io_eof(0); }]])],
+	[libmpeg3_mpeg3io_eof=1], [libmpeg3_mpeg3io_eof=0])
+	LIBS="$old_LIBS"
+	unset old_LIBS
+
+	# Check if libmpeg3 1.7 or later is used.  It is needed for
+	# several functions not present in earlier versions.
+	AC_CHECK_LIB(mpeg3, mpeg3_open)
+	AC_TRY_LINK([#include <mpeg3/libmpeg3.h>],
+		[char *s; int e; mpeg3_open(s, &e);],
+		libmpeg3_open_return_error=1, libmpeg3_open_return_error=0)
+	if test 1 = $libmpeg3_open_return_error ; then
+		echo "external libmpeg3 version with mpeg3_open(char*, int*) (present in 1.7 or later) found"
+	else
+		echo "error: external libmpeg3 without mpeg3_open(char*, int*) (before version 1.7) found"
+		echo "error: Upgrade libmpeg3 or drop the --with-external-libmpeg3 configure flag."
+		exit 1
+fi
+else
+	LIBMPEG3_CFLAGS="-I\$(top_srcdir)/libmpeg3"
+	LIBMPEG3_LIBS="\$(top_builddir)/libmpeg3/libmpeg3hv.la"
+	libmpeg3_mpeg3io_eof=1
+	AC_SUBST(LIBMPEG3_CFLAGS)
+	AC_SUBST(LIBMPEG3_LIBS)
+fi
+
+AM_CONDITIONAL(HAVE_MPEG3IO_EOF, [test "x$libmpeg3_mpeg3io_eof" = "x1"])
+AM_CONDITIONAL(HAVE_LIBMPEG3_INTERNAL, [test "x$libmpeg3_system" != "xyes"])
+############# END external libmpeg3
+
+############# CSS SUPPORT IN INTERNAL LIBMPEG3
 if test "x$enable_css" = "xyes"; then
 	CSS_CFLAGS="-DHAVE_CSS"
 fi
 AC_SUBST(CSS_CFLAGS)
 AC_ARG_ENABLE(css, 
-							AC_HELP_STRING([--disable-css], [disable support for css in libmpeg3 (default=enabled)]),,
+	AC_HELP_STRING([--disable-css], [disable support for css in internal libmpeg3 (default=enabled)]),,
 							[ enable_css=$enableval ],
 							[ enable_css=yes ])
 ############## END OF CSS SUPPORT IN LIBMPEG3
diff --git a/mpeg2enc/Makefile.am b/mpeg2enc/Makefile.am
index 30da0ef..36ef9f7 100644
--- a/mpeg2enc/Makefile.am
+++ b/mpeg2enc/Makefile.am
@@ -4,7 +4,7 @@
 
 noinst_LTLIBRARIES = libmpeg2enc.la
 
-AM_CPPFLAGS = -I$(top_srcdir)/quicktime -I$(top_srcdir)/libmpeg3
+AM_CPPFLAGS = -I$(top_srcdir)/quicktime $(LIBMPEG3_CFLAGS)
 AM_CFLAGS = $(LARGEFILE_CFLAGS)
 
 libmpeg2enc_la_SOURCES = conform.c mpeg2enc.c putseq.c putpic.c puthdr.c putmpg.c \
diff --git a/mplexlo/Makefile.am b/mplexlo/Makefile.am
index 58e76be..27c0819 100644
--- a/mplexlo/Makefile.am
+++ b/mplexlo/Makefile.am
@@ -1,7 +1,7 @@
 bin_PROGRAMS=mplexlo
 
-AM_CPPFLAGS=-I$(top_srcdir)/libmpeg3
+AM_CPPFLAGS=$(LIBMPEG3_CFLAGS)
 AM_CFLAGS = $(LARGEFILE_CFLAGS)
-mplexlo_LDADD=../libmpeg3/libmpeg3hv.la
+mplexlo_LDADD=$(LIBMPEG3_LIBS)
 mplexlo_LDFLAGS=-lm -lpthread
 mplexlo_SOURCES=mplex.c
diff --git a/quicktime/Makefile.am b/quicktime/Makefile.am
index 7c83429..6e74c88 100644
--- a/quicktime/Makefile.am
+++ b/quicktime/Makefile.am
@@ -21,7 +21,7 @@ libquicktimehv_la_LIBADD = \
 	$(FAAD_LIBS) \
 	-lfaac \
 	encore50/libencore.la \
-	$(top_builddir)/libmpeg3/libmpeg3hv.la \
+	$(LIBMPEG3_LIBS) \
 	$(LIBX264_LIBS) \
 	$(LIBDV_LIBS) \
 	-ljpeg -lpng \
@@ -129,7 +129,7 @@ noinst_HEADERS = cmodel_permutation.h \
 	
 EXTRA_DIST = docs
 
-AM_CPPFLAGS = -I$(top_srcdir)/libmpeg3
+AM_CPPFLAGS = $(LIBMPEG3_CFLAGS)
 
 pkgincludedir=$(includedir)/quicktime
 pkginclude_HEADERS=quicktime.h qtprivate.h
-- 
2.0.0.rc2

